experiment/autobumper/bumper/bumper.go

```diff
@@ -89,28 +264,126 @@ func UpdatePullRequestWithLabels(gc github.Client, org, repo, title, body, match
+		if err != nil {
+			return nil, err
+		}
+	default:
+		tagPicker = func(imageHost, imageName, currentTag string) (string, error) { return tagVersion, nil }
+	}
+
+	updateFile := func(name string) error {
+		fmt.Printf("Updating file %s\n", name)
+		if err := imagebumper.UpdateFile(tagPicker, name, filterRegexp); err != nil {
+			logrus.WithError(err).Errorf("Failed to update file %q.", name)
+			return err
+		}
+		return nil
+	}
+	updateYAMLFile := func(name string) error {
+		if strings.HasSuffix(name, ".yaml") && !isUnderPath(name, excludeConfigPaths) {
+			return updateFile(name)
+		}
+		return nil
+	}
+
+	// Updated all .yaml files under the included config paths but not under excluded config paths.
+	for _, path := range includeConfigPaths {
+		info, err := os.Stat(path)
 		if err != nil {
+			logrus.WithError(err).Errorf("Error getting file info for %q.", path)
+			return nil, err
+		}
+		if info.IsDir() {
+			err := filepath.Walk(path, func(subpath string, info os.FileInfo, err error) error {
+				return updateYAMLFile(subpath)
+			})
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			if err := updateYAMLFile(path); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	// Update the extra files in any case.
+	for _, file := range extraFiles {
+		if err := updateFile(file); err != nil {
 			return nil, err
 		}
 	}
 
-	return bumper.GetReplacements(), nil
+	return imagebumper.GetReplacements(), nil
+}
+
+func upstreamImageVersionResolver(upstreamAddress string) (func(imageHost, imageName, currentTag string) (string, error), error) {
+	version, err := parseUpstreamImageVersion(upstreamAddress)
+	if err != nil {
+		return nil, fmt.Errorf("error resolving the upstream Prow version from %q: %v", upstreamAddress, err)
+	}
+	return func(imageHost, imageName, currentTag string) (string, error) {
+		// Skip boskos images as they do not have the same image tag as other Prow components.
+		// TODO(chizhg): remove this check after all Prow instances are using boskos images not in gcr.io/k8s-prow/boskos
+		if strings.Contains(imageName, "boskos/") {
+			return currentTag, nil
+		}
+		return version, nil
+	}, nil
+}
+
+func parseUpstreamImageVersion(upstreamAddress string) (string, error) {
+	resp, err := http.Get(upstreamAddress)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("HTTP error %d (%q) fetching upstream config file", resp.StatusCode, resp.Status)
+	}
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", fmt.Errorf("error reading the response body: %v", err)
+	}
+	imagematcher := regexp.MustCompile(`(?s)^.+image:.+:(v[a-zA-Z0-9_.-]+)`)
+	res := imagematcher.FindStringSubmatch(string(body))
+	return res[1], nil
+}
+
+func isUnderPath(name string, paths []string) bool {
+	for _, p := range paths {
+		if p != "" && strings.HasPrefix(name, p) {
```

Comment:
Indenting error flow would be more like:

```golang
if p == "" || !strings.HasPrefix(name, p) {
  continue
}
return true
```

Reply:
I feel `if p != "" && strings.HasPrefix(name, p)` is more understandable, and this flow adds one more line, so I left this unchanged.